pkg/managerdriver/utils.go

```diff
@@ -422,3 +397,80 @@ func doHostGlobsMatch(hostname1 string, hostname2 string) (bool, error) {
 	}
 }
+
+func protocolStringToIRScheme(proto string) (ir.IRScheme, error) {
+	switch strings.ToUpper(proto) {
+	case "HTTP":
+		return ir.IRScheme_HTTP, nil
+	case "HTTPS":
+		return ir.IRScheme_HTTPS, nil
+	default:
+		return ir.IRScheme_HTTP, fmt.Errorf("unable to get scheme for protocol %q, expected HTTP/HTTPS", proto)
+	}
+}
+
+func getProtoForServicePort(log logr.Logger, service *corev1.Service, portName string, portNumber int32) (string, error) {
+	if service.Annotations != nil {
+		annotation := service.Annotations["k8s.ngrok.com/app-protocols"]
+		if annotation != "" {
+			log.Info("annotated app-protocols found", "annotation", annotation, "namespace", service.Namespace, "service", service.Name, "port name", portName)
+			protocolMap := map[string]string{}
+			err := json.Unmarshal([]byte(annotation), &protocolMap)
+			if err != nil {
+				return "", fmt.Errorf("could not parse protocol annotation: '%s' from: %s service: %s", annotation, service.Namespace, service.Name)
+			}
+
+			if protocol, ok := protocolMap[portName]; ok {
+				log.V(3).Info("found protocol for port name", "protocol", protocol, "namespace", service.Namespace, "service", service.Name)
+				// only allow cases through where we are sure of intent
+				switch upperProto := strings.ToUpper(protocol); upperProto {
+				case "HTTP", "HTTPS":
+					return upperProto, nil
+				default:
+					log.Error(fmt.Errorf("service uses \"k8s.ngrok.com/app-protocols\" annotation to configure protocols, but a valid entry is missing for portName :%q. Port protocol will be inferred from the port name/number", portName), "missing protocol annotation entry for service",
+						"service", fmt.Sprintf("%s.%s", service.Name, service.Namespace),
+					)
+				}
+			}
+		}
+	}
+
+	// Either there was no app-protocols annotation, or this portName was misssing one
+	switch strings.ToUpper(portName) {
+	case "HTTP":
+		log.Info("inferred protocol HTTP for service port based on port name",
+			"port name", portName,
+			"port number", portNumber,
+			"service", fmt.Sprintf("%s.%s", service.Name, service.Namespace),
+		)
+		return "HTTP", nil
+	case "HTTPS":
+		log.Info("inferred protocol HTTPS for service port based on port name",
+			"port name", portName,
+			"port number", portNumber,
+			"service", fmt.Sprintf("%s.%s", service.Name, service.Namespace),
+		)
+		return "HTTPS", nil
+	}
+
+	// Last resort, try to infer based on common HTTP/HTTPS ports
+	switch portNumber {
+	case int32(80):
+		log.Info("inferred protocol HTTP for service port based on port number",
+			"port name", portName,
+			"port number", portNumber,
+			"service", fmt.Sprintf("%s.%s", service.Name, service.Namespace),
+		)
+		return "HTTP", nil
+	case int32(443):
+		log.Info("inferred protocol HTTPS for service port based on port number",
+			"port name", portName,
+			"port number", portNumber,
+			"service", fmt.Sprintf("%s.%s", service.Name, service.Namespace),
+		)
+		return "HTTPS", nil
```

Comment:
Can we hold off on this new behavior? I'm not sure relying on the name or port number will do the right thing more times than it may do the wrong thing. In a service, someone might have named the port "HTTP", have it listening on port 443, and the scheme they actually want is `https`. 

I really like the idea, but in practice I think it might cause more problems. Maybe we punt on this for 1.0?